### ISSUE TITLE
In line 39 of cohesinKO.main.R: pairDF <- getAllCisPairs(genesGR, maxDist=10^6, minDist=0)

### ISSUE THREAD

**Issue by dpolychr** [ibn-salem/cohesinKO#1]
In line 39 of the cohesinKO.main.R, I get the following message:

pairDF <- getAllCisPairs(genesGR, maxDist=10^6, minDist=0)
Error in 1:nrow : NA/NaN argument
In addition: Warning message:
In valid.GenomicRanges.seqinfo(x, suggest.trim = TRUE) :
  GRanges object contains 295 out-of-bound ranges located on sequences chr1, chr2, chr3, chr4, chr5, chr6, chr7, chr8, chr9, chr10,
  chr11, chr12, chr13, chr14, chr15, chr16, chr17, chr18, chr19, chrX, chrY, chr1_GL456210_random, chr1_GL456221_random,
  chr4_GL456216_random, chr4_GL456350_random, chr4_JH584292_random, chr4_JH584293_random, chr4_JH584294_random,
  chr4_JH584295_random, chr5_GL456354_random, chr5_JH584296_random, chr5_JH584297_random, chr5_JH584299_random,
  chr7_GL456219_random, chrX_GL456233_random, and chrUn_JH584304. Note that only ranges located on a non-circular sequence whose
  length is not NA can be considered out-of-bound (use seqlengths() and isCircular() to get the lengths and circularity flags of
  the underlying sequences). You can use trim() to trim these ranges. See ?`trim,GenomicRanges-method` for more information.

I run trim and I got the same message.

**Comment by ibn-salem** [comment]
The code in[ `getAllCisPairs()`](https://github.com/ibn-salem/genepair/blob/master/R/genepairFunctions.R) create a temporary GRanges object in which each TSS is extended by +-10^6 bp. This result in coordinates outside the chromosomes. I think this should not be a problem here. But I have no idea on how to prevent this warning. Any suggestion is welcome:)

**Comment by dpolychr** [comment]
Thanks Jonas, I see, yes. The thing is that in the above case the pairDF object is not created at all.

I tried to put the resize functions where you extend the TSSs inside the function suppressWarnings:

getAllCisPairs <- function(genesGR, maxDist=10^6, minDist=0){
    
    # calculate overlap all possible gene pairs within maxDist bp
    posGR <- suppressWarnings(resize(genesGR, width=1, fix="start"))
    windowGR <- suppressWarnings(resize(posGR, 2*maxDist+1, fix="center"))
    
    hits = findOverlaps(posGR, windowGR, ignore.strand=TRUE)
    
    # convert hits objec to data.frame
    gp = as.data.frame(hits)
    names(gp) <- c("g1", "g2")
    
    # remove pairs with same gene
    gp <- gp[gp[,1] != gp[,2], ]
    
    # remove A-B, B-A duplicates:
    gp <- uniquePair(gp)
    
    # sort columns according to index in genesGR
    firstSmaller <- gp[,1] <= gp[,2]
    
    first <- gp[cbind(1:nrow, ifelse(firstSmaller, 1, 2))]
    second <- gp[cbind(1:nrow, ifelse(firstSmaller, 2, 1))]
    
    gp[,1] <- first
    gp[,2] <- second
    
    # sort rows according to index in genesGR
    gp <- gp[order(gp[,1], gp[,2]), ]
    
    # add distance
    gp <- addPairDist(gp, genesGR)
    
    # remove pars with distance smaller than minDist or larger than maxDist
    gp <- gp[abs(gp$dist) >= minDist,]
    
    return(gp)
    
}

The warnings are successfully suppressed, apart from the first:

 Error in 1:nrow : NA/NaN argument 

Hmmm, I have to think about it more carefully :-)

**Comment by dpolychr** [comment]
I think that the "issue" was in:

first <- gp[cbind(1:nrow, ifelse(firstSmaller, 1, 2))]
second <- gp[cbind(1:nrow, ifelse(firstSmaller, 2, 1))]

if we replace that with:

first <- gp[cbind(1:nrow(gp), ifelse(firstSmaller, 1, 2))]
second <- gp[cbind(1:nrow(gp), ifelse(firstSmaller, 2, 1))]

the function works :)

In addition, some suggestions:

a) we could use regioneR::filterChromosomes function to keep only canonical chromosomes. For instance: genesGR <-  filterChromosomes(A = genesGR, organism = "mm", chr.type = "canonical").

b) From the same package we could use regioneR::extendRegions to extend the TSSs maxDist from the start and maxDist from the end. For instance:

windowGR <- suppressWarnings(regioneR::extendRegions(posGR, extend.start = maxDist, extend.end = maxDist))

c) maybe we could even omit windowGR. For example:

getAllCisPairs <- function(genesGR, maxDist=10^6, minDist=0){
    # calculate overlap all possible gene pairs within maxDist bp
    posGR <- suppressWarnings(resize(genesGR, width=1, fix="start"))
    hits = findOverlaps(posGR, maxgap=maxDist, drop.redundant=TRUE, drop.self=TRUE)
    # convert hits objec to data.frame
    gp = as.data.frame(hits)
    names(gp) <- c("g1", "g2")
    # remove pairs with same gene
    gp <- gp[gp[,1] != gp[,2], ]
    # remove A-B, B-A duplicates:
    gp <- uniquePair(gp)
    # sort columns according to index in genesGR
    firstSmaller <- gp[,1] <= gp[,2]
    first <- gp[cbind(1:nrow(gp), ifelse(firstSmaller, 1, 2))]
    second <- gp[cbind(1:nrow(gp), ifelse(firstSmaller, 2, 1))]
    gp[,1] <- first
    gp[,2] <- second
    # sort rows according to index in genesGR
    gp <- gp[order(gp[,1], gp[,2]), ]
    # add distance
    gp <- addPairDist(gp, genesGR)
    # remove pars with distance smaller than minDist or larger than maxDist
    gp <- gp[abs(gp$dist) >= minDist,]
    return(gp)
}

I am not sure about the latter but it gives gene - pairs ... :)

**Comment by ibn-salem** [comment]
Thanks Dimitirs!
yes, your suggestions make very much sense and I included them in the latest commit (https://github.com/ibn-salem/genepair/commit/b3d2c55573856659c323a67e7f0f1ec0feec116a)
I think we also do not need the  `suppressWarnings()` anymore and the arguments `drop.redundant=TRUE` and `drop.self=TRUE` makes some other downstream steps unnecessary. I only added the `ignore.strand=TRUE` argument to include gene pairs on different strands. 
Thanks a lot!
